system/View/Table.php

```diff
@@ -295,14 +295,17 @@ public function generate($tableData = null)
 
         // Is there a table heading to display?
         if (! empty($this->heading)) {
+            preg_match('/(<)(td|th)/i', $this->template['heading_cell_start'], $valueHead);
+            $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
```

Comment:
Why do you need the `count()` evaluation? If the string matches the pattern, then `$valueHead[0]` should be the same as `$valueHead[1] . $valueHead[2]`. Isn't it?
https://3v4l.org/6rSB2

Reply:
With pattern now, i just checked `<thd>` or `<tdh>` can be accept (see [https://3v4l.org/KvjvH](https://3v4l.org/KvjvH)), i can update with `'/(<)(td|th)(>)/i'` pattern (see [https://3v4l.org/ZB3Nn](https://3v4l.org/ZB3Nn)). With new pattern i need `$valueHead[1] . $valueHead[2]`

Reply:
```php
--- a/system/View/Table.php
+++ b/system/View/Table.php
@@ -295,8 +295,8 @@ class Table

         // Is there a table heading to display?
         if (! empty($this->heading)) {
-            preg_match('/(<)(td|th)/i', $this->template['heading_cell_start'], $valueHead);
-            $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
+            preg_match('/(<)(td|th)(>)/i', $this->template['heading_cell_start'], $valueHead);
+            $strHeading = count($valueHead) === 4 ? $valueHead[1] . $valueHead[2] : null;

             $out .= $this->template['thead_open'] . $this->newline . $this->template['heading_row_start'] . $this->newline;

@@ -304,8 +304,8 @@ class Table
                 $temp = $this->template['heading_cell_start'];

                 foreach ($heading as $key => $val) {
-                    if ($key !== 'data') {
-                        $temp = $strHeading ? str_replace($strHeading, $strHeading . ' ' . $key . '="' . $val . '"', $temp) : $temp;
+                    if ($key !== 'data' && $strHeading !== null) {
+                        $temp = str_replace($strHeading, $strHeading . ' ' . $key . '="' . $val . '"', $temp);
                     }
                 }

@@ -359,8 +359,8 @@ class Table
         // Any table footing to display?
         if (! empty($this->footing)) {
             // Check if there is a th or td tag, null if not
-            preg_match('/(<)(td|th)/i', $this->template['footing_cell_start'], $valueFoot);
-            $strFooting = count($valueFoot) === 3 ? $valueFoot[1] . $valueFoot[2] : null;
+            preg_match('/(<)(td|th)(>)/i', $this->template['footing_cell_start'], $valueFoot);
+            $strFooting = count($valueFoot) === 4 ? $valueFoot[1] . $valueFoot[2] : null;

             $out .= $this->template['tfoot_open'] . $this->newline . $this->template['footing_row_start'] . $this->newline;

@@ -368,8 +368,8 @@ class Table
                 $temp = $this->template['footing_cell_start'];

                 foreach ($footing as $key => $val) {
-                    if ($key !== 'data') {
-                        $temp = $strFooting ? str_replace($strFooting, $strFooting . ' ' . $key . '="' . $val . '"', $temp) : $temp;
+                    if ($key !== 'data' && $strFooting !== null) {
+                        $temp = str_replace($strFooting, $strFooting . ' ' . $key . '="' . $val . '"', $temp);
                     }
                 }
```

Reply:
I think I didn't get my point through.

My point is in this pattern `/(<)(td|th)/i`, why use `$matches[1] . $matches[2]` when you can achieve the same with `$matches[0]`?

In https://3v4l.org/KvjvH, the `0`-key is `<td`, which is the concat of `1`-key and `2`-key.
```
array (
  0 => '<td',
  1 => '<',
  2 => 'td',
)
```

Now, on your point that the pattern will also match `<tdh>`, you can use lookahead to whitespace or `>`. Like `/(<)(td|th)(?=\h|>)/i`. See https://3v4l.org/2fm0O

Reply:
Thank you suggest `/(<)(td|th)(?=\h|>)/i` pattern, when I access $valueHead[0] directly return error

```
There were 2 errors:

1) CodeIgniter\View\TableTest::testGenerateInvalidHeadingFooting
Undefined offset: 0

D:\Project\laragon\www\ci4\system\View\Table.php:299
D:\Project\laragon\www\ci4\tests\system\View\TableTest.php:568
phpvfscomposer://D:\Project\laragon\www\ci4\vendor\phpunit\phpunit\phpunit:97

2) CodeIgniter\View\TableTest::testGenerateInvalidHeadingFootingHTML
Undefined offset: 0

D:\Project\laragon\www\ci4\system\View\Table.php:299
D:\Project\laragon\www\ci4\tests\system\View\TableTest.php:617
phpvfscomposer://D:\Project\laragon\www\ci4\vendor\phpunit\phpunit\phpunit:97

ERRORS!
Tests: 26, Assertions: 78, Errors: 2.
```

That's why I use `$valueHead[1] . $valueHead[2]`
![image](https://user-images.githubusercontent.com/97607754/190092776-7fda7546-f0f0-4e8e-9efc-f4037362cf3d.png)

Reply:
No error